File: exercicios/regex.py
import re

>>> resultado = re.search(r'\d\d\w','11a22b33c')
`

>>> resultado = re.search(r'(\d\d)\w','11a22b33c')
>>> resultado.group()
11a
>>> resultado.group(1)
11
>>>resultado.start()
0 
>>>resultado.end()
3

>>> resultados = re.finditer(r'(\d\d)\w','11a22b33c')
>>> for resultado in resultados:
...     print "%s com grupo %s [%s,%s]" % (resultado.group(), resultado.group(1),resultado.start(), resultado.end())
... 
11a com grupo 11 [0,3]
22b com grupo 22 [3,6]
33c com grupo 33 [6,9]

regex = re.compile(r'(\d\d)\w')
resultados = regex.finditer('11a22b33c')




# ============================================================================================
							# método findall

>>> import re

>>> regex = re.compile(r'(\d\d)(\w)')
>>> alvo = '11a22b33c'


>>> resultado = re.findall(regex, alvo)

>>> print resultado
[('11', 'a'), ('22', 'b'), ('33', 'c')]

>>> resultado[0]
('11', 'a')
>>> resultado[1]
('22', 'b')
>>> resultado[2]
('33', 'c')


>>> for grupo in resultado:
...     print grupo
... 
('11', 'a')
('22', 'b')
('33', 'c')



>>> for grupo in resultado:
...     print grupo[0] + grupo[1]
... 
11a
22b
33c

# ======================================================================================

					 	# Alterando o separador de uma data

>>> alvo = '2007-12-31'

>>> import re
>>> regex = '\s-\s'
>>> novotexto = ': '
>>> alura = 'Alura - Regex'
>>> resultado = re.sub(regex, novotexto, alura)



>>> print resultado
Alura: Regex


# Se queremos trocar todos os hífens por uma /, podemos fazer:

>>> import re
>>> regex = '-'
>>> novotexto = '/'
>>> alvo = '2007-12-31'
>>> resultado = re.sub(regex, novotexto, alvo)


>>> print resultado
2007/12/31

File: exercicios/regex.java
// No mundo Java, a classe principal para trabalhar com expressões regulares se chama Pattern e faz parte do pacote java.util.regex.

// Para criar uma regex, devemos usar o método estático compile(regexString), da classe Pattern, por exemplo:

Pattern pattern = Pattern.compile("(\\d\\d)(\\w)");
// Repare que foi necessário escapar o símbolo \. Outra coisa para notar é que o Java não possui uma sintaxe literal para a definição de regex.

// Uma vez declarado o pattern, devemos criar um objeto Matcher, baseado na string de entrada:

Matcher matcher = pattern.matcher("11a22b33c");
// Esse matcher possui, por sua vez, os métodos para acessar o match, o grupo, index, etc, mas antes de tudo devemos chamar o método find, que verifica se realmente há um match devolvendo true ou false:

boolean encontrou = matcher.find();
// A partir daí podemos pegar os valores do match:

if(encontrou) {
    String match = matcher.group();
    String group1 = matcher.group(1);
    String group2 = matcher.group(2);

    int start = matcher.start();
    int end = matcher.end();

    System.out.printf("%s | %s |  %s [%d,%d] %n", match, group1, group2, start, end);
}
// Sabendo disso, podemos colocar o nosso código dentro de um laço while para mostrar todos os resultados:

public class TesteRegex {

    private static Pattern pattern = Pattern.compile("(\\d\\d)(\\w)");

    public static void main(String[] args) {

        Matcher matcher = pattern.matcher("11a22b33c");
        while(matcher.find()) {
            String match = matcher.group();
            String group1 = matcher.group(1);
            String group2 = matcher.group(2);

            int start = matcher.start();
            int end = matcher.end();

            System.out.printf("%s | %s |  %s [%d,%d] %n", match, group1, group2, start, end);
        }
    }
}
// As linguagens funcionam bem parecidas trabalhando com expressões regulares e o Java não é diferente.